### ISSUE TITLE
250704 Machine Learning Basic_김시진

### ISSUE THREAD

**Issue by kimsijin33** [AIBootcamp14/wiki-rider#10]
📅 학습 날짜  
*2025-07-04*

## 📌 키워드 및 요약 정리
키워드:
강화학습, 손실함수, MAE, MSE, RMSE, 교차 엔트로피, 정량평가 지표, 정성평가, 검증용 데이터셋, 모델의 복잡도, Regularization, 과적합, 혼돈행렬, 정밀도, 재현율, F1 스코어, ROC Curve, AUC, 최적화 지표, 충족 지표

용어 정의: 
**강화학습**은 모델이 게임과 같은 환경에서 상호작용하게 설정, 특정 행동을 했을 때 리워드 부여하여 목표를 달성하게 유도해서 학습하는 ML의 한 분야. 
정성평가(Qualitative Evaluation)는 모델에 들어가는 입력값과 그에 따른 모델의 출력값을 사람이 직접 확인해 보면서, 모델의 성능이 어떤지 판단하는 것. 
정량평가(Quantitative Evaluation) 지표는 모델의 성능이 좋은지 수치로 나타내는 평가지표(Evaluation Metric), 함수를 설정해 두고, 이 지표를 활용해 모델 성능의 좋고 나쁨을 판단하는 것. 

핵심원리: 
**손실함수**(Loss Function)는 딥러닝 모델의 성능을 측정하고 최적화하는 데 중요한 역할을 하는 함수.
**혼돈행렬**(Confusion Matrix)은 이진 분류 문제에서 모델의 예측치와 실제값에 따른 샘플의 수를 표 형태로 나타낸 것.

실습: 
정밀도 계산, 재현율 계산

## 📘 개념 정리 

데이터셋은 사람의 판단을 모아놓은 것이며 의도하지 않았더라도 편견이나 선입견이 반영될 수 있음

### 1. 모델 평가의 구성 요소

정량평가 지표는 모델이 정답을 얼마나 맞혔는지를 수치로 나타내는 지표임
손실함수(Loss Function)는 딥러닝 모델의 성능을 측정하고 최적화하는 데 중요한 역할을 하는 함수
손실함수는 실제값과 예측값의 차이로 계산되므로 성능지표로 볼수 있음.

MAE(Mean Absolute Error) 평균 절대 오차
오차에 절대값을 취한 것의 평균, 작은 오류 민감

평균 절대 오차의 수식 
<img width="223" height="73" alt="Image" src="https://github.com/user-attachments/assets/ea5a3138-28b9-4c2c-85ee-126597194407" />
yi 는 실제 값, yi hat은 예측 값, n은 데이터의 총 개수임

MSE(Mean Squared Error) 평균 제곱 오차
오차를 제곱한 것의 평균, 이상치 유무 큰 영향

평균 제곱 오차의 수식
<img width="227" height="70" alt="Image" src="https://github.com/user-attachments/assets/de73dd8d-6620-434e-b24d-bfeffe7aefe2" />
yi는 실제 값을 의미, yi hat은 예측 값을 의미 하고, n은 데이터의 총 개수임

RMSE(Root Mean Squared Error)
MSE의 제곱근으로, MSE를 사용하면 오차가 원래의 수치와 다른 단위로 표시된다는 점을 보완함.

교차 엔트로피 손실(Cross Entropy Loss)
교차 엔트로피 손실함수는 모델의 예측 확률 분포와 실제 레이블(정답) 간의 차이를 측정하는데 사용

교차 엔트로피 손실의 수식
<img width="381" height="75" alt="Image" src="https://github.com/user-attachments/assets/1e628313-65ad-4ce5-b221-8445c5f4dc67" />

yi는 실제 레이블 (0 또는 1)을 의미함
yi hat은 예측 확률(0 과 1 사이)을 의미함
n은 데이터의 총 개수임

평가 데이터셋의 구성
Train, Test 데이터셋 분할 - 9:1 학습용 / 평가용
검증용 데이터셋(Validation Dataset) - 실제로 8:1:1 학습용 / 검증용 / 평가용 세개로 사용하는 것이 흔함.

모델의 복잡도 (Complicity, Capacity) 
모델이 더 복잡하고 어려운 데이터까지 학습할 수 있는 용량.
일반적으로 파라미터의 수가 많을수록 (딥러닝의 경우 레이어 및 뉴런 수가 많을수록), 복잡도가 커짐

Regularization
과적합 문제의 방지를 위해, 모델의 복잡도가 필요 이상으로 커지자 않도록 제한을 두는 방법론

과적합(Overfitting)
모델이 학습데이터의 패턴에만 과하게 학습되어, 손심함수 값은 0에 수렴하지만 평가 데이터셋 등 새로운 데이터에 대해서는 전혀 성능을 내지 못함

혼돈행렬(Confusion Matrix)
이진 분류 문제에서 모델의 예측치와 실제값에 따른 샘플의 수를 표 형태로 나타낸 것.
예측이 맞았는지 틀렸는지를 나타내는 (True|False)와 결과가 양성인지 음성인지를 나타내는 (Positive|Negative)를 붙여 표시함.

혼돈행렬 그림
<img width="467" height="235" alt="Image" src="https://github.com/user-attachments/assets/6d037431-91fe-4c39-9d5e-199bb4818f74" />
초록색 음영은 예측이 맞은 것이고, 빨간색 음영은 틀린 것임

정밀도(Precision)
모델이 양성으로 예측한 샘플중 정답(실제로 양성)의 비율 
식: TP / (TP + FP)

재현율(Recall)
모델이 양성으로 정확하게 분류한 샘플의 비율 
식: TP / (TP + FN)

정밀도와 재현율 그림
<img width="394" height="738" alt="Image" src="https://github.com/user-attachments/assets/51f76d6d-a887-44e4-bbb4-5a1ba1b6190d" />

F1 스코어(F1 Score)
Recall과 Precision을 동시에 비교하기 위해 조화평균(Harmonic Mean)을 취한 것. H-mean 이라고도 함.

실습:
어떤 질병의 검사 결과 |   |   |  
-- | -- | -- | --
총 100,000명 중 | 총 인구 | 검사 양성 | 검사 음성
실제 질병 걸림 | 1,000 | 950 (TP) | 50 (FN)
실제 질병 걸리지 않음 | 99,000 | 4,950 (FP) | 94,050 (TN)

Recall = TP / TP + FN = 950 / 950 + 50 = 0.95
Precision = TP / TP + FP = 950 / 950 + 4950 = 0.1610169492
F1Score = 2 / Recall의 -1승 + Precison의 -1승 = 2 / 1.052631579 + 6.210526314 = 0.2753623189

ROC(Receiver Operating Characteristic) Curve
음성 /양성을 판단하는 기준값(Threshold)의 변화에 따라, 전체에서 TP, FP의 비율이 변화하는 추세를 나타낸 곡선

AUC(Area Under the ROC Curve)
ROC Curve의 아래쪽 영역의 넓이로 측정되며, 정밀도와 재현율을 모두 중요하게 고려하는 정량평가 지표.

[컴퓨팅 자원과 관련된 정량평가 지표들] 
인퍼런스 시간(Latency)
모델이 샘플 한개, 또는 여러개를 한번에 처리하는데 걸리는 시간

처리율(Throughput)
모델이 단위시간당 처리할 수 있는 샘플의 수

메모리 사용량(Memory Usage)
(특히 GPU에서) 모델을 인퍼런스할 때 점유하는 메모리(VRAM)의 크기

### 2. 정량평가의 체계적인 활용

정성평가, 정량평가 모두를 적절히 활용해야 함.

연구를 진행한다면 완전히 동일한 평가지표를 사용해야하며, 기존 연구들의 측정방식을 참고

서비스를 만든다면 서비스 자체의 목표와 밀접한 연관이 있는 평가지표를 선정, 예를 들어 유투브 추천시스템은 시청시간을 이용해 평가

이진분류 문제
일반적 정확도 사용 
데이터 불균형(Imbalance)문제가 있는 경우, 정밀도, 재현율, F1 스코어 등을 추가 사용

예시: 
초기 유투브는 조회수 기반 인기 영상 판단 2012년부터 영상의 시청시간 기반의 평가 시작
초기에는 조회수가 급락하는 등 부작용을 겪었지만 결국 전반적인 영상 추천의 품질을 올릴 수 있었다고 함.
2016년 부터는 영상 시청시간과 유저 설문을 결합한 Valued Watchtime 지표를 사용

## ❓ 어려웠던 부분  
- 혼돈행렬의 정밀도와 재현율을 TP, TN, FP, FN에 따라 계산하는 것이 조금 헷갈립니다. 

## 💬 질문 또는 토론 유도  
- 인공지능의 세부 분야에 따라 성능평가에 주로 어떤 평가지표를 사용하는지 궁금합니다.
예를들어 자연어처리 분야의 LLM, 비전, 시계열 데이터 등

## 🔗 참고 자료 (선택)  
- 정밀도와 재현율 위키문서 - https://en.wikipedia.org/wiki/Precision_and_recall
- 손실함수 - https://wikidocs.net/277027
- 혼돈행렬 - https://en.wikipedia.org/wiki/Confusion_matrix

**Comment by Young-SeokOh** [comment]
시진 선생님, 안녕하세요. 제가 피드백 날짜를 착각하여 늦었습니다. 미안합니다. 우선 평가 지표들을 다양한 표현 방식으로 정리를 잘해주셨습니다 :) 질문 주신 것과 관련해서는 NLP, CV 등 다양한 AI 도메인마다 그 지표가 다릅니다. 왜냐하면 일반적으로도 시각과 언어는 평가 방식이 달라야하기 때문입니다. 그리고 머신러닝에서 배운 지표와 관련해서도 분야마다 중요시 여기는 지표가 다릅니다. 이 부분은 시진 선생님께서 관심가지고 계신 분야에서 어떤 지표를 쓰고 있는지 직접 찾아보시고 공부하시면 많이 도움이 되실 것입니다. 각자마다 관심분야가 다르기에 이와같이 답변드리는 것이 옳을 듯 합니다.

평가지표와 관련해서 정리된 논문을 추천 논문으로 올려드립니다. 평안한 하루되세요 :)

[s41598-024-56706-x.pdf](https://github.com/user-attachments/files/21102191/s41598-024-56706-x.pdf)
